pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+// service resources (like external load balancers) in sync with the registry.
+func NewServiceController(
+	cloud cloudprovider.Interface,
+	kubeClient client.Interface,
+	clusterName string) *ServiceController {
+	return &ServiceController{
+		cloud:       cloud,
+		kubeClient:  kubeClient,
+		clusterName: clusterName,
+		serviceMap:  make(map[string]*api.Service),
+	}
+}
+
+// Run starts a background goroutine that watches for changes to services that
+// have (or had) externalLoadBalancers=true and ensures that they have external
+// load balancers created and deleted appropriately.
+func (s *ServiceController) Run() {
+	if err := s.init(); err != nil {
+		glog.Errorf("Failed to initialized ServiceController: %v", err)
+		return
+	}
+
+	// We have to make this check beecause the ListWatch that we use in
+	// WatchServices requires Client functions that aren't in the interface
+	// for some reason.
+	if _, ok := s.kubeClient.(*client.Client); !ok {
+		glog.Error("ServiceController only works with real Client objects, but was passed something else satisfying the client Interface.")
+		return
+	}
+
+	go s.watchServices()
+}
+
+func (s *ServiceController) init() error {
+	if s.cloud == nil {
+		return fmt.Errorf("ServiceController should not be run without a cloudprovider.")
+	}
+
+	balancer, ok := s.cloud.TCPLoadBalancer()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support external TCP load balancers.")
+	}
+	s.balancer = balancer
+
+	zones, ok := s.cloud.Zones()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support zone enumeration, which is required for creating external load balancers.")
+	}
+	zone, err := zones.GetZone()
+	if err != nil {
+		return fmt.Errorf("failed to get zone from cloud provider, will not be able to create external load balancers: %v", err)
+	}
+	s.zone = zone
+	return nil
+}
+
+func (s *ServiceController) watchServices() {
+	// Get the currently existing set of services and then all future creates
+	// and updates of services.
+	// TODO: Add a compressor that intelligently squashes together updates.
+	keyLister := cache.KeyListerFunc(func() []string { return s.listKeys() })
+	serviceQueue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, keyLister)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*client.Client), "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, 0).Run()
+	// TODO: Add proper retries rather than just re-adding to the queue?
+	for {
+		newItem := serviceQueue.Pop()
+		deltas, ok := newItem.(cache.Deltas)
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", newItem)
+		}
+		delta := deltas.Newest()
+		if delta == nil {
+			glog.Errorf("Received nil delta from watcher queue.")
+			continue
+		}
+		err := s.processDelta(delta)
+		if err != nil {
+			// Add the failed service back to the queue so we'll retry it.
+			glog.Error(err)
+			time.Sleep(5 * time.Second)
```

Comment:
Having an arbitrarily long sleep like this buried deep in your code like this can be dangerous.   Let me go through the rest of your code before proposing alternatives.